internal/controller/resourceset_controller.go

```diff
@@ -0,0 +1,643 @@
+// for the resources to become ready.
+// It returns an error if the apply operation fails, otherwise
+// it returns the sha256 digest of the applied resources.
+func (r *ResourceSetReconciler) apply(ctx context.Context,
+	obj *fluxcdv1.ResourceSet,
+	objects []*unstructured.Unstructured) (string, error) {
+	log := ctrl.LoggerFrom(ctx)
+	var changeSetLog strings.Builder
+
+	// Create a snapshot of the current inventory.
+	oldInventory := inventory.New()
+	if obj.Status.Inventory != nil {
+		obj.Status.Inventory.DeepCopyInto(oldInventory)
+	}
+
+	// Configure the Kubernetes client for impersonation.
+	impersonation := runtimeClient.NewImpersonator(
+		r.Client,
+		r.StatusPoller,
+		polling.Options{},
+		nil,
+		runtimeClient.KubeConfigOptions{},
+		r.DefaultServiceAccount,
+		obj.Spec.ServiceAccountName,
+		obj.GetNamespace(),
+	)
+
+	// Create the Kubernetes client that runs under impersonation.
+	kubeClient, statusPoller, err := impersonation.GetClient(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to build kube client: %w", err)
+	}
+
+	// Create a resource manager to reconcile the resources.
+	resourceManager := ssa.NewResourceManager(kubeClient, statusPoller, ssa.Owner{
+		Field: r.StatusManager,
+		Group: fmt.Sprintf("resourceset.%s", fluxcdv1.GroupVersion.Group),
+	})
+	resourceManager.SetOwnerLabels(objects, obj.GetName(), obj.GetNamespace())
+
+	if err := normalize.UnstructuredList(objects); err != nil {
+		return "", err
+	}
+
+	if cm := obj.Spec.CommonMetadata; cm != nil {
+		ssautil.SetCommonMetadata(objects, cm.Labels, cm.Annotations)
+	}
+
+	// Compute the sha256 digest of the resources.
+	data, err := ssautil.ObjectsToYAML(objects)
+	if err != nil {
+		return "", fmt.Errorf("failed to convert objects to YAML: %w", err)
+	}
+	applySetDigest := digest.FromString(data).String()
+
+	applyOpts := ssa.DefaultApplyOptions()
+	applyOpts.Cleanup = ssa.ApplyCleanupOptions{
+		// Remove the kubectl and helm annotations.
+		Annotations: []string{
+			corev1.LastAppliedConfigAnnotation,
+			"meta.helm.sh/release-name",
+			"meta.helm.sh/release-namespace",
+		},
+		// Remove the flux labels set at bootstrap.
+		Labels: []string{
+			"kustomize.toolkit.fluxcd.io/name",
+			"kustomize.toolkit.fluxcd.io/namespace",
+		},
+		// Take ownership of the Flux resources if they
+		// were previously managed by other tools.
+		FieldManagers: []ssa.FieldManager{
+			{
+				Name:          "flux",
+				OperationType: metav1.ManagedFieldsOperationApply,
+			},
+			{
+				Name:          "kustomize-controller",
+				OperationType: metav1.ManagedFieldsOperationApply,
+			},
+			{
+				Name:          "helm",
+				OperationType: metav1.ManagedFieldsOperationUpdate,
+			},
+			{
+				Name:          "kubectl",
+				OperationType: metav1.ManagedFieldsOperationUpdate,
+			},
+		},
+	}
+
+	resultSet := ssa.NewChangeSet()
+
+	// Apply the resources to the cluster.
+	changeSet, err := resourceManager.ApplyAllStaged(ctx, objects, applyOpts)
```

Comment:
I see that kustomize-controller instead of using this function breaks down the manifests in the stages:

* CRDs + Namespaces
* Classes
* Everything else

While `ApplyAllStaged` merges classes with everything else.

What's the use for kustomize-controller to do that? Do we need it here too?

I see that `FluxInstance` also uses `ApplyAllStaged` but it doesn't deploy any classes so it's trivially fine in its case.